### DISCUSSION TITLE
Is there a simple solution to watch for changes of entities in a `withEntities()` signal store?

### DISCUSSION THREAD

**Discussion by ptandler** [ngrx/platform#4571] [Q&A]
Scenario:

I use a `signalStore(withEntities<Entity>() /*...*/)`.

I have a use case when I want to trigger some other updates whenever an entity in the store is added, updated or removed.

I'd like to use an `effect()` - but I miss an simple & elegant solution to find out about what changed since last call of the effect function.

I even don't get the previous store state so I can compare myself.

Any suggestion?

Ideally, I'd like to have something like

```ts
watchEntities(store, (updated: EntityId[], added: EntityId[], removed: EntityId[]) => {
  // called as effect whenever things change in the store so I can
  // do whatever I need to keep other things in sync with these entities
};
```

AI suggested to do something like:

```ts
toObservable(store).pipe(
  startWith(null),
  pairwise()
).subscribe(([prev, current]) => {
  if (prev === null) return; // Skip initial emission
  // ...and an awful lot of code to detect the changes...
```

So I'd say it's simple to augment my modification methods to trigger the actions.

**Comment by rainerhahnekamp** [comment]
Well, there is nothing really built-in for `withEntities`. You could use `rxMethod` or listen to the Slices in order to find out what has really changed.

Do you have some use cases? Maybe, this could be a required feature.

**Comment by ptandler** [comment]
Our use case is a store that creates a "manager" object for each entity that is added to the store and ensure a 1-to-1 mapping with identical IDs for both. The idea is that when we receive plain JSON objects from the backend, we can put them to the store - and for each one create another object that provides computed stuff, to e.g. apply defaults (but also more complex computations) without altering the original objects and update when the entity changes.

Or, is there already a way to add computed signals and methods for each entity in a store?

Another idea for a use cases of watchEntities() could be the handling of external resources that are linked to each entity, requests that should be made to fetch details or similar.

**Reply by PhilippMDoerner** [reply]
Why not just use `withComputed` ? 
It seems to me that you must define a function that transforms A (your JSON Input) to B (your derived Output) either way and you must define for each JSON Input which of your functions applies anyway. A simple `withComputed` here seems like the most maintainable approach to me. 

There is an alternative where you can generate that `withComputed` call automatically for all keys on your store (and by having specific suffixes on the fieldname you can prevent this being applied to non-entity fields) etc., but that means writing a custom store feature that is very generic. 
Particularly when you start deriving fields from other fields that can be somewhat tricky  (see [this discussion](https://github.com/ngrx/platform/discussions/4566) for reference) because you need to provide TS the correct Type Information *and* define how your input type gets turned into Output types.

Particularly in scenarios where you derive both field names and their types from input fields, you're often working with typecasting (or at least I ended up going down that route) and still need to provide some input-data in the beginning to even be able to derive the correct output-type. 
Thus you may end up not really saving yourself much typing.

So given you're already having those transformation functions somewhere anyway, I think a simple withComputed like this would be easier:.
``` ts
...
withComputed(store => {
  const transformService = inject(MyTransformationService) //Maybe you need DI or sth
  return {
    valADerived: computed(() => deriveA(store.valA())
    ...
  }
})
...
```
Note: I am by no means an expert, so anything rainer says supersedes anything I write on the matter. I just played around and managed to build a couple custom store features that work essentially the same way `withMethods` works (you put in a factory function that gets a state parameter, which adds a bunch of things to the store). Particularly getting the types right proved to be on the trickier side of things for me.

**Reply by ptandler** [reply]
In our case these "manager" objects are passed around in the app, so I need encapsulated objects and it wouldn't be practical to access the store in those places.

I can, of course, use withComputed() to create the manager objects, but here as well I need to make sure that I don't create new instances of the manager objects when the entities change. Rather I have to keep the instances and ensure that their signals are re-computed.

I have a solution that works. But it's not nice :-)

**Reply by rainerhahnekamp** [reply]
I would have to try it out on my own, but have you ever thought of a nested computed? The outer computed only tracks the array of entities. It only changes if a new entity gets added or another is removed. The computed exposes then a record where the key is again the id of the entity and the value would be a computed wrapping the entity with your special transformation.

**Reply by ptandler** [reply]
I tried this. The part I don't like is that I need some external non-signal place to store the previous result of my computed wrappers, to be able to re-use them in the re-computation.

So, currently I have a non-reactive implementation that relies on a single method modifying the store:

```ts
import {
  patchState,
  signalStoreFeature,
  type,
  withMethods,
} from '@ngrx/signals';
import type { EntityId } from '@ngrx/signals/entities';
import {
  entityConfig,
  removeAllEntities,
  removeEntity,
  setEntities,
  withEntities,
} from '@ngrx/signals/entities';
import { isEqual } from 'lodash-es';

/**
 * create a manager for each entity in the store
 */
export function withManagedEntities<
  K extends EntityId,
  V extends { id: K },
  M extends { item(): V; setItem(item: V): void },
>(createManager: (item: V) => M) {
  const managerConfig = entityConfig({
    entity: type<M>(),
    collection: 'manager',
    selectId: (item) => item.item().id,
  });

  return signalStoreFeature(
    // main store for the entities
    withEntities<V>(),
    // additional store for the managers
    withEntities(managerConfig),
    // extend store to also manage the managers
    withMethods((store) => ({
      initialize(entities: V[]) {
        // ensure we have no previous state left
        patchState(store, removeAllEntities());
        // and use saveAll method, as this might be overwritten
        this.saveAll(entities);
      },

      allIds(): K[] {
        return store.ids() as K[]; // well, we know that ids are always of type K
        // would like to avoid the cast here -> https://github.com/ngrx/platform/issues/4570
      },

      getById(id: K | undefined | null): V | undefined {
        return id ? store.entityMap()[id] : undefined;
      },

      getManagerById(id: K | undefined | null): M | undefined {
        if (!id) return undefined;
        const map = store.managerEntityMap();
        return id ? map[id] : undefined;
      },
      /**
       * also add manager for each entity
       */
      saveAll(values: ReadonlyArray<V>): K[] {
        const updatedKeys: K[] = [];
        const updatedValues: V[] = [];
        const updatedManagers: M[] = [];

        for (const value of values) {
          const existingEntity = this.getById(value.id);
          if (!existingEntity || !isEqual(existingEntity, value)) {
            updatedKeys.push(value.id);
            updatedValues.push(value);

            const existingManager = this.getManagerById(value.id);
            if (existingManager) {
              existingManager.setItem(value);
              updatedManagers.push(existingManager);
            } else {
              updatedManagers.push(createManager(value));
            }
          }
        }

        patchState(
          store,
          setEntities(updatedValues),
          setEntities(updatedManagers, managerConfig),
        );
        return updatedKeys;
      },

      /**
       * also remove manager from store
       */
      remove(id: string | undefined | null) {
        if (id) {
          patchState(store, removeEntity(id), removeEntity(id, managerConfig));
        }
      },
    })),
  );
}

```
